features/mbedtls/targets/TARGET_NXP/aes_alt.c

```diff
@@ -0,0 +1,609 @@
+/******************************************************************************/
+#if (defined(FSL_FEATURE_SOC_CAU3_COUNT) && (FSL_FEATURE_SOC_CAU3_COUNT > 0)) || (defined(MBEDTLS_FREESCALE_DCP_AES))
+static const void *s_mbedtlsCtx[4] = {0};
+
+static void crypto_attach_ctx_to_key_slot(const void *ctx, uint8_t keySlot)
+{
+    s_mbedtlsCtx[keySlot] = ctx;
+}
+
+static void crypto_detach_ctx_from_key_slot(const void *ctx)
+{
+    for (int i = 0; i < 4; i++)
+    {
+        if (ctx == s_mbedtlsCtx[i])
+        {
+            s_mbedtlsCtx[i] = NULL;
+            break;
+        }
+    }
+}
+
+static bool crypto_key_is_loaded(const void *ctx)
+{
+    bool ret = false;
+    for (int i = 0; i < 4; i++)
+    {
+        if (ctx == s_mbedtlsCtx[i])
+        {
+            ret = true;
+            break;
+        }
+    }
+    return ret;
+}
+#endif
+
+#if defined(MBEDTLS_AES_ALT)
+
+/* Implementation that should never be optimized out by the compiler */
+static void mbedtls_zeroize( void *v, size_t n ) {
+    volatile unsigned char *p = (unsigned char*)v; while( n-- ) *p++ = 0;
+}
+
+void mbedtls_aes_init( mbedtls_aes_context *ctx )
+{
+    memset( ctx, 0, sizeof( mbedtls_aes_context ) );
+#if defined(FSL_FEATURE_SOC_LTC_COUNT) && (FSL_FEATURE_SOC_LTC_COUNT > 0)
+    LTC_Init(LTC_INSTANCE);
+#endif
+#if defined(FSL_FEATURE_SOC_DCP_COUNT) && (FSL_FEATURE_SOC_DCP_COUNT > 0)
+    /* Initialize DCP */
+    dcp_config_t dcpConfig;
+
+    DCP_GetDefaultConfig(&dcpConfig);
+    DCP_Init(DCP, &dcpConfig);
+#endif
+}
+
+void mbedtls_aes_free( mbedtls_aes_context *ctx )
+{
+    if( ctx == NULL )
+        return;
+
+    mbedtls_zeroize( ctx, sizeof( mbedtls_aes_context ) );
+#if defined(FSL_FEATURE_SOC_LTC_COUNT) && (FSL_FEATURE_SOC_LTC_COUNT > 0)
+    LTC_Deinit(LTC_INSTANCE);
+#endif
+#if defined(FSL_FEATURE_SOC_DCP_COUNT) && (FSL_FEATURE_SOC_DCP_COUNT > 0)
+    DCP_Deinit(DCP);
+#endif
+}
+
+/*
+ * AES key schedule (encryption)
+ */
+int mbedtls_aes_setkey_enc(mbedtls_aes_context *ctx, const unsigned char *key, unsigned int keybits)
+{
+    uint32_t *RK;
+
+#ifdef MBEDTLS_AES_ALT_NO_192
+    if (keybits == 192u)
+    {
+        return (MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE);
+    }
+#endif
+
+#ifdef MBEDTLS_AES_ALT_NO_256
+    if (keybits == 256u)
+    {
+        return (MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE);
+    }
+#endif
+
+#if defined(MBEDTLS_FREESCALE_LTC_AES) || defined(MBEDTLS_FREESCALE_LPC_AES) || defined(MBEDTLS_FREESCALE_CAU3_AES) || \
+    defined(MBEDTLS_FREESCALE_CAAM_AES) || defined(MBEDTLS_FREESCALE_DCP_AES)
+    const unsigned char *key_tmp = key;
+    ctx->rk = RK = ctx->buf;
+    memcpy(RK, key_tmp, keybits / 8);
+
+#if defined(MBEDTLS_FREESCALE_CAU3_AES) || defined(MBEDTLS_FREESCALE_DCP_AES)
+    crypto_detach_ctx_from_key_slot(ctx);
+#endif /* MBEDTLS_FREESCALE_CAU3_AES || MBEDTLS_FREESCALE_DCP_AES */
+
+    switch (keybits)
+    { /* Set keysize in bytes.*/
+        case 128:
+            ctx->nr = 16;
```

Comment:
Please add a C comment explaining why `nr` should be set like this on this hardware. Same for mbedtls_aes_setkey_dec().

Reply:
This still needs a C comment, please.

Reply:
I have added this comment.